### ISSUE TITLE
Document forking behaviour

### ISSUE THREAD

**Issue by hanche** [elves/elvish#1391]
As far as I can tell by experimenting a bit, the sections of a pipeline, as well as background jobs, are run in the same process as the foreground. I guess in separate threads, or goroutines? (I know nothing about goroutines, other than the name.) Obviously, external commands are run in spawned subcommands. This is all about elvish code.

This should be documented, since users coming from other shells might reasonably expect these to run in a forked process instead. In particular, that means they have to watch for reuse of variables and various concurrency issues.

Speaking of which, maybe it could be useful to have some form of semaphore available, in case someone decides to do some serious parallel processing with cooperating threads.

**Comment by krader1961** [comment]
> This should be documented, since users coming from other shells might reasonably expect these to run in a forked process instead.

I agree it might be a good idea to include a statement similar to bash's "Functions are executed in the context  of  the  current  shell;  no  new process  is created to interpret them". However, I don't understand why you claim new Elvish users will "expect these to run in a forked process." Consider functions in POSIX shells like bash or ksh. From the bash documentation:

```
FUNCTIONS
       A  shell  function, defined as described above under SHELL GRAMMAR, stores a series
       of commands for later execution.  When the name of a shell function is  used  as  a
       simple  command  name,  the  list of commands associated with that function name is
       executed.  Functions are executed in the context  of  the  current  shell;  no  new
       process  is  created to interpret them (contrast this with the execution of a shell
       script).  When a function is executed, the arguments to  the  function  become  the
       positional  parameters during its execution.  The special parameter # is updated to
       reflect the change.  Special parameter 0 is unchanged.  The first  element  of  the
       FUNCNAME  variable is set to the name of the function while the function is execut-
       ing.
```

That quite clearly says that bash functions are run in the context of the current shell process and therefore can modify its state. The Elvish behavior is fundamentally the same. Consider this interactive bash example:

```
bash-5.1$ export WTF=wtf
bash-5.1$ env | grep WTF
WTF=wtf
bash-5.1$ function x { WTF=hello; echo $$; }
bash-5.1$ echo $$
72021
bash-5.1$ x
72021
bash-5.1$ env | grep WTF
WTF=hello
```

Notice that function `x` modified the environment of the current shell process because it ran in the context of that process.

So I'm wondering, @hanche, if I am missing some context for why you opened this issue.

**Comment by krader1961** [comment]
See also issue #1003 and some related issues you can find by searching for issues that include terms like "mutex".

**Comment by hanche** [comment]
@krader1961 Well, you did not run any of your shell commands _in a pipeline_. It makes a difference. Also, the bash manual says  about the `$` variable:

> Expands  to  the process ID of the shell.  In a () subshell, it expands to the process ID of the current shell, not the subshell.

so relying on an external program seems better. And indeed, what the manual says for a () subshell also seems to apply to commands in a pipeline:

```
> cat ./ppid
#!/bin/bash
echo $@ $PPID
> export A=wtf
> export B=wtf
> function a { A=a; echo a '$$' $$; ./ppid a pid; }
> function b { B=b; echo b '$$' $$; ./ppid b pid; }
> a >&2 | b
a $$ 18798
b $$ 18798
a pid 18818
b pid 18819
> echo $$ $A $B
18798 wtf wtf
```

Contrast with elvish:

```
⬥ var A B = outer outer
⬥ fn a { set A = a; echo 'a $pid '$pid; ./ppid a pid }
⬥ fn b { set B = b; echo 'b $pid $pid'; ./ppid b pid }
⬥ a >&2 | b
b $pid $pid
a $pid 18345
b pid 18345
a pid 18345
⬥ echo $pid $A $B
18345 a b
```

On the off chance that the bash version is relevant, since the Apple-supplied bash on macOS is ancient (3.5.7), I installed bash from macports (and got version 5.1.8) and ran the experiment there, with the same result.

**Comment by krader1961** [comment]
Regarding `$$`: Yeah, I continually forget that bash (and probably every POSIX shell) has that weird quirk. Bash has `$BASHPID` for sane behavior while ksh only has the broken by design `$$`.

Regarding functions in pipelines. I think you're sort of making my point. The fact that using a function in a pipeline changes its side-effects is wrong by design. Note too that whether the function (or other code block) is run in a subshell depends on the shell and whether the function is the first or last command in the pipeline. This is a well known gotcha of the Bourne shell (*/bin/sh*). This doesn't work:

```
> x="false"
> echo true | while read x ; do x=$x; done
> echo $x
false
```

The Korn shell (`ksh`) fixed that design mistake. However, note that it means the function only affects the state of the current shell process if it is the last command in the pipeline:

```
KSH PROMPT:11: export WTF=hello
KSH PROMPT:12: x | true
KSH PROMPT:13: echo $WTF
hello
KSH PROMPT:14: true | x
KSH PROMPT:15: echo $WTF
goodbye
```

But `bash` gets it wrong in that it doesn't matter where the function is placed in the pipeline:

```
bash-5.1$ function x { WTF=goodbye; }
bash-5.1$ export WTF=hello
bash-5.1$ x | true
bash-5.1$ echo $WTF
hello
bash-5.1$ true | x
bash-5.1$ echo $WTF
hello
bash-5.1$ x
bash-5.1$ echo $WTF
goodbye
```

Having said all that I can't see any harm in documenting the current behavior since I can't imagine why it would ever be changed to match legacy shells like `bash` (bad) or `ksh` (less bad).

**Comment by hanche** [comment]
> Regarding functions in pipelines. I think you're sort of making my point. 

Ah, if your point was that the elvish way is better, I wholeheartedly agree. I am pretty sure the old behaviour stems from the fact that separate threads of execution in a single process wasn't there from the beginning of unix, far from it. The available mechanism was `fork()`, so that is what one used. The alternative, possible perhaps but tedious, would be to hack multiple threads of execution of shell commands without any support from the OS or the language itself.
